From: Justus Winter <justus@g10code.com>
Date: Wed, 1 Mar 2017 17:47:47 +0100
Subject: common,tools: Always escape newlines when escaping data.

* common/stringhelp.c (do_percent_escape): Always escape newlines.
* tools/gpgconf-comp.c (gc_percent_escape): Likewise.
--
Newlines always pose a problem for a line-based communication format.

GnuPG-bug-id: 2387
Signed-off-by: Justus Winter <justus@g10code.com>
(cherry picked from commit e064c75b08a523f738108428fe0c417a46e66238)
---
 common/stringhelp.c  | 10 +++++++++-
 tools/gpgconf-comp.c |  7 +++++++
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/common/stringhelp.c b/common/stringhelp.c
index dea2212c4..0abfa3d3a 100644
--- a/common/stringhelp.c
+++ b/common/stringhelp.c
@@ -1052,7 +1052,8 @@ do_percent_escape (const char *str, const char *extra, int die)
     return NULL;
 
   for (i=j=0; str[i]; i++)
-    if (str[i] == ':' || str[i] == '%' || (extra && strchr (extra, str[i])))
+    if (str[i] == ':' || str[i] == '%' || str[i] == '\n'
+        || (extra && strchr (extra, str[i])))
       j++;
   if (die)
     ptr = xmalloc (i + 2 * j + 1);
@@ -1077,6 +1078,13 @@ do_percent_escape (const char *str, const char *extra, int die)
 	  ptr[i++] = '2';
 	  ptr[i++] = '5';
 	}
+      else if (*str == '\n')
+	{
+	  /* The newline is problematic in a line-based format.  */
+	  ptr[i++] = '%';
+	  ptr[i++] = '0';
+	  ptr[i++] = 'a';
+	}
       else if (extra && strchr (extra, *str))
         {
 	  ptr[i++] = '%';
diff --git a/tools/gpgconf-comp.c b/tools/gpgconf-comp.c
index 530c1287f..9358e2efa 100644
--- a/tools/gpgconf-comp.c
+++ b/tools/gpgconf-comp.c
@@ -1490,6 +1490,13 @@ gc_percent_escape (const char *src)
 	  *(dst++) = '2';
 	  *(dst++) = 'c';
 	}
+      else if (*src == '\n')
+	{
+	  /* The newline is problematic in a line-based format.  */
+	  *(dst++) = '%';
+	  *(dst++) = '0';
+	  *(dst++) = 'a';
+	}
       else
 	*(dst++) = *(src);
       src++;
